README.md
# httpcache

httpcache provides utilities for HTTP Caching.

> [!WARNING]
> This package is still under development.


CHANGELOG.md
# Changelog

## [v0.0.1](https://github.com/k1LoW/httpcache/commits/v0.0.1) - 2023-12-13


httpcache_test.go
package httpcache

import (
	"fmt"
	"net/http"
	"testing"
)

func TestHandlerToClientDo(t *testing.T) {
	tests := []struct {
		h    http.Handler
		req  *http.Request
		want *http.Response
	}{
		{
			http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				w.WriteHeader(http.StatusOK)
				w.Header().Set("Content-Type", "text/plain")
				_, _ = w.Write([]byte("OK"))
			}),
			&http.Request{},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Content-Type": []string{"text/plain"},
				},
			},
		},
	}
	for i, tt := range tests {
		tt := tt
		t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
			t.Parallel()
			got, err := HandlerToClientDo(tt.h)(tt.req)
			if err != nil {
				t.Fatal(err)
			}
			if got.StatusCode != tt.want.StatusCode {
				t.Errorf("StatusCode: got %d, want %d", got.StatusCode, tt.want.StatusCode)
			}
			if got.Header.Get("Content-Type") != tt.want.Header.Get("Content-Type") {
				t.Errorf("Content-Type: got %s, want %s", got.Header.Get("Content-Type"), tt.want.Header.Get("Content-Type"))
			}
		})
	}
}


rfc9111/shared_test.go
package rfc9111

import (
	"net/http"
	"net/url"
	"testing"
	"time"

	"github.com/google/go-cmp/cmp"
)

func TestShared_Storable(t *testing.T) {
	now := time.Date(2024, 12, 13, 14, 15, 16, 00, time.UTC)

	tests := []struct {
		name        string
		req         *http.Request
		res         *http.Response
		wantOK      bool
		wantExpires time.Time
	}{
		{
			"GET 200 Cache-Control: s-maxage=10 -> +10s",
			&http.Request{
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Cache-Control": []string{"s-maxage=10"},
				},
			},
			true,
			time.Date(2024, 12, 13, 14, 15, 26, 00, time.UTC),
		},
		{
			"GET 200 Cache-Control: max-age=15 -> +15s",
			&http.Request{
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Cache-Control": []string{"max-age=15"},
				},
			},
			true,
			time.Date(2024, 12, 13, 14, 15, 31, 00, time.UTC),
		},
		{
			"GET 200 Expires: 2024-12-13 14:15:20",
			&http.Request{
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Expires": []string{"Mon, 13 Dec 2024 14:15:20 GMT"},
				},
			},
			true,
			time.Date(2024, 12, 13, 14, 15, 20, 00, time.UTC),
		},
		{
			"GET 200 Expires: 2024-12-13 14:15:20, Date: 2024-12-13 13:15:20",
			&http.Request{
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Expires": []string{"Mon, 13 Dec 2024 14:15:20 GMT"},
					"Date":    []string{"Mon, 13 Dec 2024 13:15:20 GMT"},
				},
			},
			true,
			time.Date(2024, 12, 13, 15, 15, 16, 00, time.UTC),
		},
		{
			"GET 200 Last-Modified: 2024-12-13 14:15:10, Date: 2024-12-13 14:15:20 -> +1s",
			&http.Request{
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{"Mon, 13 Dec 2024 14:15:10 GMT"},
					"Date":          []string{"Mon, 13 Dec 2024 14:15:20 GMT"},
				},
			},
			true,
			time.Date(2024, 12, 13, 14, 15, 21, 00, time.UTC),
		},
		{
			"GET 200 Last-Modified: 2024-12-13 14:15:06 -> +1s",
			&http.Request{
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{"Mon, 13 Dec 2024 14:15:06 GMT"},
				},
			},
			true,
			time.Date(2024, 12, 13, 14, 15, 17, 00, time.UTC),
		},
		{
			"GET 500 Last-Modified: 2024-12-13 14:15:06 -> No Store",
			&http.Request{
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusInternalServerError,
				Header: http.Header{
					"Last-Modified": []string{"Mon, 13 Dec 2024 14:15:06 GMT"},
				},
			},
			false,
			time.Time{},
		},
		{
			"GET 200 -> No Store",
			&http.Request{
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Date": []string{"Mon, 13 Dec 2024 14:15:10 GMT"},
				},
			},
			false,
			time.Time{},
		},
		{
			"UNUNDERSTOODMETHOD 200 -> No Store",
			&http.Request{
				Method: "UNUNDERSTOODMETHOD",
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Cache-Control": []string{"max-age=15"},
				},
			},
			false,
			time.Time{},
		},
		{
			"GET 100 -> No Store",
			&http.Request{
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusContinue,
				Header: http.Header{
					"Cache-Control": []string{"max-age=15"},
				},
			},
			false,
			time.Time{},
		},
		{
			"GET 206 -> No Store",
			&http.Request{
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusPartialContent,
				Header: http.Header{
					"Cache-Control": []string{"max-age=15"},
				},
			},
			false,
			time.Time{},
		},
		{
			"GET 200 Cache-Control: no-store -> No Store",
			&http.Request{
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Cache-Control": []string{"no-store"},
				},
			},
			false,
			time.Time{},
		},
		{
			"GET 200 Cache-Control: private -> No Store",
			&http.Request{
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Cache-Control": []string{"private"},
				},
			},
			false,
			time.Time{},
		},
		{
			"GET 500 Cache-Control: public, Last-Modified 2024-12-13 14:15:06 -> +1s",
			&http.Request{
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{"Mon, 13 Dec 2024 14:15:06 GMT"},
					"Cache-Control": []string{"public"},
				},
			},
			true,
			time.Date(2024, 12, 13, 14, 15, 17, 00, time.UTC),
		},
		{
			"GET 500 Cache-Control: public -> No Store",
			&http.Request{
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Cache-Control": []string{"public"},
				},
			},
			false,
			time.Time{},
		},
		{
			"GET Authorization: XXX 200 Cache-Control: max-age=15 -> No Store",
			&http.Request{
				Method: http.MethodGet,
				Header: http.Header{
					"Authorization": []string{"XXX"},
				},
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Cache-Control": []string{"max-age=15"},
				},
			},
			false,
			time.Time{},
		},
	}
	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()
			s, err := NewShared()
			if err != nil {
				t.Errorf("Shared.Storable() error = %v", err)
				return
			}
			gotOK, gotExpires := s.Storable(tt.req, tt.res, now)
			if gotOK != tt.wantOK {
				t.Errorf("Shared.Storable() gotOK = %v, want %v", gotOK, tt.wantOK)
			}
			if !gotExpires.Equal(tt.wantExpires) {
				t.Errorf("Shared.Storable() gotExpires = %v, want %v", gotExpires, tt.wantExpires)
			}
		})
	}
}

func TestShared_Handle(t *testing.T) {
	// 2024-12-13 14:15:16
	now := time.Date(2024, 12, 13, 14, 15, 16, 00, time.UTC)
	fresh := now.Add(15 * time.Second)
	stale := now.Add(-15 * time.Second)

	endpoint, err := url.Parse("https://example.com/api/v1/path/to/resource")
	if err != nil {
		t.Fatal(err)
	}
	other, err := url.Parse("https://example.com/api/v2/path/to/resource")
	if err != nil {
		t.Fatal(err)
	}

	origin200res := &http.Response{
		StatusCode: http.StatusOK,
		Header:     http.Header{},
	}
	origin304res := &http.Response{
		StatusCode: http.StatusNotModified,
		Header:     http.Header{},
	}
	do200 := func(req *http.Request) (*http.Response, error) {
		return origin200res, nil
	}
	do304 := func(req *http.Request) (*http.Response, error) {
		return origin304res, nil
	}
	tests := []struct {
		name          string
		req           *http.Request
		cachedReq     *http.Request
		cachedRes     *http.Response
		do            func(req *http.Request) (*http.Response, error)
		wantCacheUsed bool
		wantRes       *http.Response
	}{
		{
			"No cached request/response (nil)",
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
			},
			nil,
			nil,
			do200,
			false,
			origin200res,
		},
		{
			"Use origin response (defferent URL)",
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
			},
			&http.Request{
				URL:    other,
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{fresh.Format(http.TimeFormat)},
				},
			},
			do200,
			false,
			origin200res,
		},
		{
			"Use origin response (defferent method)",
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
			},
			&http.Request{
				URL:    endpoint,
				Method: http.MethodHead,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{fresh.Format(http.TimeFormat)},
				},
			},
			do200,
			false,
			origin200res,
		},
		{
			"Use origin response (Vary: *)",
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
			},
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Vary":          []string{"*"},
					"Last-Modified": []string{fresh.Format(http.TimeFormat)},
				},
			},
			do200,
			false,
			origin200res,
		},
		{
			"Use flesh cached response",
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
			},
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{fresh.Format(http.TimeFormat)},
					"Date":          []string{fresh.Format(http.TimeFormat)},
				},
			},
			do200,
			true,
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{fresh.Format(http.TimeFormat)},
					"Date":          []string{fresh.Format(http.TimeFormat)},
				},
			},
		},
		{
			"Use stale cached response",
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
			},
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{stale.Format(http.TimeFormat)},
					"Date":          []string{stale.Format(http.TimeFormat)},
				},
			},
			do200,
			true,
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{stale.Format(http.TimeFormat)},
					"Date":          []string{stale.Format(http.TimeFormat)},
				},
			},
		},
		{
			"Use origin response (Cache-Control: max-stale=10)",
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
				Header: http.Header{
					"Cache-Control": []string{"max-stale=10"},
				},
			},
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{stale.Format(http.TimeFormat)},
					"Date":          []string{stale.Format(http.TimeFormat)},
				},
			},
			do200,
			false,
			origin200res,
		},
		{
			"Use stale cached response (Cache-Control: max-stale=20)",
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
				Header: http.Header{
					"Cache-Control": []string{"max-stale=20"},
				},
			},
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{stale.Format(http.TimeFormat)},
					"Date":          []string{stale.Format(http.TimeFormat)},
				},
			},
			do200,
			true,
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{stale.Format(http.TimeFormat)},
					"Date":          []string{stale.Format(http.TimeFormat)},
				},
			},
		},
		{
			"Use flesh cached response (Vary: Content-Type, User-Agent)",
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
				Header: http.Header{
					"User-Agent":   []string{"test"},
					"Content-Type": []string{"application/json"},
				},
			},
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
				Header: http.Header{
					"Content-Type": []string{"application/json"},
					"User-Agent":   []string{"test"},
				},
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{fresh.Format(http.TimeFormat)},
					"Date":          []string{fresh.Format(http.TimeFormat)},
					"Vary":          []string{"content-type, user-agent"},
				},
			},
			do200,
			true,
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{fresh.Format(http.TimeFormat)},
					"Date":          []string{fresh.Format(http.TimeFormat)},
					"Vary":          []string{"content-type, user-agent"},
				},
			},
		},
		{
			"Use origin response (Vary: Content-Type, User-Agent)",
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
				Header: http.Header{
					"User-Agent":   []string{"test2"},
					"Content-Type": []string{"application/json"},
				},
			},
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
				Header: http.Header{
					"Content-Type": []string{"application/json"},
					"User-Agent":   []string{"test"},
				},
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{fresh.Format(http.TimeFormat)},
					"Vary":          []string{"content-type, user-agent"},
				},
			},
			do200,
			false,
			origin200res,
		},
		{
			"Use origin response (Cache-Control: no-cache)",
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
				Header: http.Header{},
			},
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
				Header: http.Header{},
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{fresh.Format(http.TimeFormat)},
					"Cache-Control": []string{"no-cache"},
				},
			},
			do200,
			false,
			origin200res,
		},
		{
			"Validate and use origin response",
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
				Header: http.Header{},
			},
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
				Header: http.Header{},
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{stale.Format(http.TimeFormat)},
					"Date":          []string{stale.Format(http.TimeFormat)},
					"Cache-Control": []string{"must-revalidate"},
				},
			},
			do200,
			false,
			origin200res,
		},
		{
			"Validate and use cached response",
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
				Header: http.Header{},
			},
			&http.Request{
				URL:    endpoint,
				Method: http.MethodGet,
				Header: http.Header{},
			},
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{stale.Format(http.TimeFormat)},
					"Date":          []string{stale.Format(http.TimeFormat)},
					"Cache-Control": []string{"must-revalidate"},
				},
			},
			do304,
			true,
			&http.Response{
				StatusCode: http.StatusOK,
				Header: http.Header{
					"Last-Modified": []string{stale.Format(http.TimeFormat)},
					"Date":          []string{stale.Format(http.TimeFormat)},
					"Cache-Control": []string{"must-revalidate"},
				},
			},
		},
	}
	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			//t.Parallel()
			s, err := NewShared()
			if err != nil {
				t.Errorf("Shared.Handle() error = %v", err)
				return
			}
			gotCacheUsed, gotRes, err := s.Handle(tt.req, tt.cachedReq, tt.cachedRes, tt.do, now)
			if err != nil {
				t.Errorf("Shared.Handle() error = %v", err)
				return
			}
			if gotCacheUsed != tt.wantCacheUsed {
				t.Errorf("Shared.Handle() gotCacheUsed = %v, want %v", gotCacheUsed, tt.wantCacheUsed)
			}
			if diff := cmp.Diff(gotRes, tt.wantRes); diff != "" {
				t.Errorf("Shared.Handle() gotRes != tt.wantRes:\n%s", diff)
			}
		})
	}
}


rfc9111/shared.go
package rfc9111

import (
	"net/http"
	"strings"
	"time"

	"github.com/k1LoW/httpcache"
)

var _ httpcache.Handler = (*Shared)(nil)

// Shared is a shared cache that implements RFC 9111.
// The following features are not implemented
// - Private cache.
// - Request directives.
type Shared struct {
	understoodMethods                 []string
	understoodStatusCodes             []int
	heuristicallyCacheableStatusCodes []int
	heuristicExpirationRatio          float64
}

// SharedOption is an option for Shared.
type SharedOption func(*Shared) error

// HeuristicExpirationRatio sets the heuristic expiration ratio.
func HeuristicExpirationRatio(ratio float64) SharedOption {
	return func(s *Shared) error {
		s.heuristicExpirationRatio = ratio
		return nil
	}
}

// NewShared returns a new Shared cache handler.
func NewShared(opts ...SharedOption) (*Shared, error) {
	s := &Shared{
		heuristicExpirationRatio: defaultHeuristicExpirationRatio,
	}

	um := make([]string, len(defaultUnderstoodMethods))
	_ = copy(um, defaultUnderstoodMethods)
	s.understoodMethods = um

	us := make([]int, len(defaultUnderstoodStatusCodes))
	_ = copy(us, defaultUnderstoodStatusCodes)
	s.understoodStatusCodes = us

	hs := make([]int, len(defaultHeuristicallyCacheableStatusCodes))
	_ = copy(hs, defaultHeuristicallyCacheableStatusCodes)
	s.heuristicallyCacheableStatusCodes = hs

	for _, opt := range opts {
		if err := opt(s); err != nil {
			return nil, err
		}
	}

	return s, nil
}

// Storable returns true if the response is storable in the cache.
func (s *Shared) Storable(req *http.Request, res *http.Response, now time.Time) (bool, time.Time) {
	// 3. Storing Responses in Caches (https://www.rfc-editor.org/rfc/rfc9111#section-3)
	// - the request method is understood by the cache;
	if !contains(req.Method, s.understoodMethods) {
		return false, time.Time{}
	}

	// - the response status code is final (see https://www.rfc-editor.org/rfc/rfc9110#section-15);
	if contains(res.StatusCode, []int{
		http.StatusContinue,
		http.StatusSwitchingProtocols,
		http.StatusProcessing,
		http.StatusEarlyHints,
	}) {
		return false, time.Time{}
	}

	rescc := ParseResponseCacheControlHeader(res.Header.Values("Cache-Control"))

	// - if the response status code is 206 or 304, or the must-understand cache directive (see https://www.rfc-editor.org/rfc/rfc9111#section-5.2.2.3) is present: the cache understands the response status code;
	if contains(res.StatusCode, []int{
		http.StatusPartialContent,
		http.StatusNotModified,
	}) || (rescc.MustUnderstand && !contains(res.StatusCode, s.understoodStatusCodes)) {
		return false, time.Time{}
	}

	// - the no-store cache directive is not present in the response (see https://www.rfc-editor.org/rfc/rfc9111#section-5.2.2.5);
	if rescc.NoStore {
		return false, time.Time{}
	}

	// - if the cache is shared: the private response directive is either not present or allows a shared cache to store a modified response; see https://www.rfc-editor.org/rfc/rfc9111#section-5.2.2.7);
	if rescc.Private {
		return false, time.Time{}
	}

	// - if the cache is shared: the Authorization header field is not present in the request (see https://www.rfc-editor.org/rfc/rfc9111#section-11.6.2 of [HTTP]) or a response directive is present that explicitly allows shared caching (see https://www.rfc-editor.org/rfc/rfc9111#section-3.5);
	// In this specification, the following response directives have such an effect: must-revalidate (https://www.rfc-editor.org/rfc/rfc9111#section-5.2.2.2), public (https://www.rfc-editor.org/rfc/rfc9111#section-5.2.2.9), and s-maxage (https://www.rfc-editor.org/rfc/rfc9111#section-5.2.2.10).
	if req.Header.Get("Authorization") != "" && !rescc.MustRevalidate && !rescc.Public && rescc.SMaxAge == nil {
		return false, time.Time{}
	}

	expires := CalclateExpires(rescc, res.Header, s.heuristicExpirationRatio, now)
	if expires.Sub(now) <= 0 {
		return false, time.Time{}
	}

	// - the response contains at least one of the following:

	//   * a public response directive (see https://www.rfc-editor.org/rfc/rfc9111#section-5.2.2.9);
	if rescc.Public {
		return true, expires
	}
	//   * a private response directive, if the cache is not shared (see https://www.rfc-editor.org/rfc/rfc9111#section-5.2.2.7);
	// THE CACHE IS SHARED

	//   * an Expires header field (see https://www.rfc-editor.org/rfc/rfc9111#section-5.3);
	if res.Header.Get("Expires") != "" {
		return true, expires
	}
	//   * a max-age response directive (see https://www.rfc-editor.org/rfc/rfc9111#section-5.2.2.1);
	if rescc.MaxAge != nil {
		return true, expires
	}
	//   * if the cache is shared: an s-maxage response directive (see https://www.rfc-editor.org/rfc/rfc9111#section-5.2.2.10);
	if rescc.SMaxAge != nil {
		return true, expires
	}
	//   * a cache extension that allows it to be cached (see https://www.rfc-editor.org/rfc/rfc9111#section-5.2.3); or
	// NOT IMPLEMENTED

	//   * a status code that is defined as heuristically cacheable (see https://www.rfc-editor.org/rfc/rfc9111#section-4.2.2).
	if contains(res.StatusCode, s.heuristicallyCacheableStatusCodes) {
		return true, expires
	}

	return false, time.Time{}
}

func (s *Shared) Handle(req *http.Request, cachedReq *http.Request, cachedRes *http.Response, do func(*http.Request) (*http.Response, error), now time.Time) (bool, *http.Response, error) {
	if cachedReq == nil || cachedRes == nil {
		res, err := do(req)
		return false, res, err
	}

	// 4. Constructing Responses from Caches
	// When presented with a request, a cache MUST NOT reuse a stored response unless:

	// - the presented target URI (Section 7.1 of [HTTP]) and that of the stored response match, and
	if req.URL.String() != cachedReq.URL.String() {
		res, err := do(req)
		return false, res, err
	}

	// - the request method associated with the stored response allows it to be used for the presented request, and
	if req.Method != cachedReq.Method { // FIXME: more strictly
		res, err := do(req)
		return false, res, err
	}

	// - request header fields nominated by the stored response (if any) match those presented (see https://www.rfc-editor.org/rfc/rfc9111#section-4.1)
	if v := cachedRes.Header.Values("Vary"); len(v) != 0 {
		vary := strings.Join(v, ",")
		if strings.Contains(vary, "*") {
			res, err := do(req)
			return false, res, err
		}
		for _, h := range strings.Split(vary, ",") {
			h = strings.TrimSpace(h)
			if req.Header.Get(h) != cachedReq.Header.Get(h) { // FIXME: more strictly
				res, err := do(req)
				return false, res, err
			}
		}
	}

	// - the stored response does not contain the no-cache directive (https://www.rfc-editor.org/rfc/rfc9111#section-5.2.2.4), unless it is successfully validated (https://www.rfc-editor.org/rfc/rfc9111#section-4.3), and
	rescc := ParseResponseCacheControlHeader(cachedRes.Header.Values("Cache-Control"))

	if rescc.NoCache {
		res, err := do(req)
		return false, res, err
	}

	expires := CalclateExpires(rescc, cachedRes.Header, s.heuristicExpirationRatio, now)

	// - the stored response is one of the following:
	//   * fresh (see https://www.rfc-editor.org/rfc/rfc9111#section-4.2), or
	if expires.Sub(now) > 0 {
		return true, cachedRes, nil
	}

	//   * allowed to be served stale (see https://www.rfc-editor.org/rfc/rfc9111#section-4.2.4), or
	if !rescc.NoCache && !rescc.MustRevalidate && rescc.SMaxAge == nil && !rescc.ProxyRevalidate {
		//     > A cache MUST NOT generate a stale response if it is prohibited by an explicit in-protocol directive (e.g., by a no-cache response directive, a must-revalidate response directive, or an applicable s-maxage or proxy-revalidate response directive; see https://www.rfc-editor.org/rfc/rfc9111#section-5.2.2).
		reqcc := ParseRequestCacheControlHeader(req.Header.Values("Cache-Control"))
		//     > A cache MUST NOT generate a stale response unless it is disconnected or doing so is explicitly permitted by the client or origin server (e.g., by the max-stale request directive in https://www.rfc-editor.org/rfc/rfc9111#section-5.2.1, extension directives such as those defined in [RFC5861], or configuration in accordance with an out-of-band contract).
		if reqcc.MaxStale == nil {
			// If no value is assigned to max-stale, then the client will accept a stale response of any age (ref https://www.rfc-editor.org/rfc/rfc9111#section-5.2.1.2).
			return true, cachedRes, nil
		}
		if expires.Add(time.Duration(*reqcc.MaxStale)*time.Second).Sub(now) > 0 {
			return true, cachedRes, nil
		}
	}

	//   * successfully validated (see https://www.rfc-editor.org/rfc/rfc9111#section-4.3).
	if req.Method == http.MethodGet || req.Method == http.MethodHead {
		if cachedRes.Header.Get("ETag") != "" {
			req.Header.Set("If-None-Match", cachedRes.Header.Get("ETag"))
		}
		if cachedRes.Header.Get("Last-Modified") != "" {
			req.Header.Set("If-Modified-Since", cachedRes.Header.Get("Last-Modified"))
		}
		res, err := do(req)
		if err != nil {
			return false, res, err
		}
		if res.StatusCode == http.StatusNotModified {
			return true, cachedRes, nil
		}
		return false, res, nil
	}

	res, err := do(req)
	return false, res, err
}

func CalclateExpires(d *ResponseDirectives, header http.Header, heuristicExpirationRatio float64, now time.Time) time.Time {
	// 	4.2.1. Calculating Freshness Lifetime
	// A cache can calculate the freshness lifetime (denoted as freshness_lifetime) of a response by evaluating the following rules and using the first match:

	// - If the cache is shared and the s-maxage response directive (https://www.rfc-editor.org/rfc/rfc9111#section-5.2.2.10) is present, use its value, or
	if d.SMaxAge != nil {
		return now.Add(time.Duration(*d.SMaxAge) * time.Second)
	}
	// - If the max-age response directive (https://www.rfc-editor.org/rfc/rfc9111#section-5.2.2.1) is present, use its value, or
	if d.MaxAge != nil {
		return now.Add(time.Duration(*d.MaxAge) * time.Second)
	}
	if header.Get("Expires") != "" {
		// - If the Expires response header field (https://www.rfc-editor.org/rfc/rfc9111#section-5.3) is present, use its value minus the value of the Date response header field
		et, err := http.ParseTime(header.Get("Expires"))
		if err == nil {
			if header.Get("Date") != "" {
				dt, err := http.ParseTime(header.Get("Date"))
				if err == nil {
					return now.Add(et.Sub(dt))
				}
			} else {
				// (using the time the message was received if it is not present, as per Section 6.6.1 of [HTTP])
				return et // == return now.Add(et.Sub(now))
			}
		}
	}
	// Otherwise, no explicit expiration time is present in the response. A heuristic freshness lifetime might be applicable; see https://www.rfc-editor.org/rfc/rfc9111#section-4.2.2.
	if header.Get("Last-Modified") != "" {
		lt, err := http.ParseTime(header.Get("Last-Modified"))
		if err == nil {
			// If the response has a Last-Modified header field (Section 8.8.2 of [HTTP]), caches are encouraged to use a heuristic expiration value that is no more than some fraction of the interval since that time. A typical setting of this fraction might be 10%.
			if header.Get("Date") != "" {
				dt, err := http.ParseTime(header.Get("Date"))
				if err == nil {
					return dt.Add(time.Duration(float64(dt.Sub(lt)) * heuristicExpirationRatio))
				}
			} else {
				return now.Add(time.Duration(float64(now.Sub(lt)) * heuristicExpirationRatio))
			}
		}
	}

	return now
}

func contains[T comparable](v T, vv []T) bool {
	for _, vvv := range vv {
		if vvv == v {
			return true
		}
	}
	return false
}
